File: Packages/TeamTasksKit/Sources/TeamTasksContract/Scenarios/ReadScenarios.swift
import Foundation
import TeamTasksCore

// PostgREST reads (docs/CONTRACTS.md §4.3): group tasks, my tasks, assignments since.
extension ContractScenarios {
    static let readScenarios: [ContractScenario] = [
        ContractScenario("reads.groupTasksAreScoped") { harness in
            let alice = try await harness.user("Alice")
            let first = try await alice.makeGroup("Un")
            let second = try await alice.makeGroup("Deux")
            let a1 = try await alice.makeTask(in: first.id)
            let a2 = try await alice.makeTask(in: first.id, assignees: [alice])
            let b1 = try await alice.makeTask(in: second.id)
            _ = try await alice.tasks.setStatus(taskId: a2.id, status: .done)
            for includeOldDone in [false, true] {
                let firstTasks = try await alice.tasks.tasks(groupId: first.id, includeOldDone: includeOldDone)
                try Verify.equal(Set(firstTasks.map(\.id)), [a1.id, a2.id], "tasks of the first group (\(includeOldDone))")
                let secondTasks = try await alice.tasks.tasks(groupId: second.id, includeOldDone: includeOldDone)
                try Verify.equal(secondTasks.map(\.id), [b1.id], "tasks of the second group (\(includeOldDone))")
                for task in firstTasks + secondTasks {
                    try Verify.equal(task.myAssignedAt, nil, "group tasks have no myAssignedAt")
                    try Verify.equal(task.groupName, nil, "group tasks have no groupName")
                }
            }
        },

        ContractScenario("reads.myTasks") { harness in
            let alice = try await harness.user("Alice")
            let bob = try await harness.user("Bob")
            let first = try await alice.makeGroup("Un", joinedBy: [bob])
            let second = try await alice.makeGroup("Deux", joinedBy: [bob])
            let todo = try await alice.makeTask(in: first.id, assignees: [bob], dueAt: Fixed.dueA)
            let done = try await alice.makeTask(in: second.id, assignees: [bob, alice])
            _ = try await bob.tasks.setStatus(taskId: done.id, status: .done)
            let notMine = try await alice.makeTask(in: first.id, assignees: [alice])
            let selfAssigned = try await bob.makeTask(in: first.id, assignees: [bob])

            let open = try await bob.tasks.myTasks(includeDone: false)
            try Verify.equal(Set(open.map(\.id)), [todo.id, selfAssigned.id], "open tasks assigned to bob")
            let all = try await bob.tasks.myTasks(includeDone: true)
            try Verify.equal(Set(all.map(\.id)), [todo.id, done.id, selfAssigned.id], "all tasks assigned to bob")
            try Verify.that(!all.contains { $0.id == notMine.id }, "tasks not assigned to bob are excluded")

            let names = [first.id: first.name, second.id: second.name]
            for item in all {
                try Verify.equal(item.groupName, names[item.groupId], "groupName of \(item.title)")
                try Verify.that(item.myAssignedAt != nil, "myAssignedAt of \(item.title)")
                let plain = try await bob.tasks.task(id: item.id)
                try Verify.equal(item.withoutPersonalFields, plain, "myTasks item \(item.title) matches task(id:)")
            }
            let doneItem = try Verify.unwrap(all.first { $0.id == done.id }, "done task in myTasks")
            try Verify.equal(doneItem.assigneeIds, sortedIDs([alice, bob]), "all assignees are listed, sorted")
            try Verify.equal(doneItem.status, .done, "status of the done task")
            let aliceOpen = try await alice.tasks.myTasks(includeDone: false)
            try Verify.equal(Set(aliceOpen.map(\.id)), [notMine.id], "alice's open tasks")
        },

        ContractScenario("reads.assignmentsSince") { harness in
            let alice = try await harness.user("Alice")
            let bob = try await harness.user("Bob")
            let group = try await alice.makeGroup(joinedBy: [bob])
            let first = try await alice.makeTask(in: group.id, "Premiere", assignees: [bob], dueAt: Fixed.dueA)
            let own = try await bob.makeTask(in: group.id, assignees: [bob])
            let second = try await alice.makeTask(in: group.id, "Seconde", assignees: [alice, bob])

            let events = try await bob.tasks.assignments(since: Fixed.longAgo)
            try Verify.equal(events.map(\.taskId), [first.id, second.id], "assignments by others, oldest first")
            try Verify.that(!events.contains { $0.taskId == own.id }, "self-assignments are excluded")
            let event = events[0]
            try Verify.equal(event.groupId, group.id, "event groupId")
            try Verify.equal(event.taskTitle, first.title, "event taskTitle")
            try Verify.equal(event.groupName, group.name, "event groupName")
            try Verify.equal(event.assignedBy, alice.id, "event assignedBy")
            try Verify.equal(event.dueAt, Fixed.dueA, "event dueAt")
            try Verify.that(events[0].assignedAt < events[1].assignedAt, "events ordered by assignedAt")

            let later = try await bob.tasks.assignments(since: events[0].assignedAt)
            try Verify.equal(later.map(\.taskId), [second.id], "since is exclusive")
            let none = try await bob.tasks.assignments(since: events[1].assignedAt)
            try Verify.that(none.isEmpty, "nothing after the last assignment, got \(none)")
            let aliceEvents = try await alice.tasks.assignments(since: Fixed.longAgo)
            try Verify.that(aliceEvents.isEmpty, "alice only assigned herself, got \(aliceEvents)")
        },
    ]
}

File: android/contract/src/main/kotlin/io/github/notkanaa/equipe/contract/ReadScenarios.kt
package io.github.notkanaa.equipe.contract

import io.github.notkanaa.equipe.core.TaskStatus

// PostgREST reads (docs/CONTRACTS.md §4.3): group tasks, my tasks, assignments since.
internal val readScenarios: List<ContractScenario> = listOf(
    ContractScenario("reads.groupTasksAreScoped") { harness ->
        val alice = harness.user("Alice")
        val first = alice.makeGroup("Un")
        val second = alice.makeGroup("Deux")
        val a1 = alice.makeTask(first.id)
        val a2 = alice.makeTask(first.id, assignees = listOf(alice))
        val b1 = alice.makeTask(second.id)
        alice.tasks.setStatus(a2.id, TaskStatus.DONE)
        for (includeOldDone in listOf(false, true)) {
            val firstTasks = alice.tasks.tasks(first.id, includeOldDone)
            Verify.equal(
                firstTasks.map { it.id }.toSet(), setOf(a1.id, a2.id), "tasks of the first group ($includeOldDone)",
            )
            val secondTasks = alice.tasks.tasks(second.id, includeOldDone)
            Verify.equal(secondTasks.map { it.id }, listOf(b1.id), "tasks of the second group ($includeOldDone)")
            for (task in firstTasks + secondTasks) {
                Verify.equal(task.myAssignedAt, null, "group tasks have no myAssignedAt")
                Verify.equal(task.myAssignedBy, null, "group tasks have no myAssignedBy")
                Verify.equal(task.groupName, null, "group tasks have no groupName")
            }
        }
    },

    ContractScenario("reads.myTasks") { harness ->
        val alice = harness.user("Alice")
        val bob = harness.user("Bob")
        val first = alice.makeGroup("Un", joinedBy = listOf(bob))
        val second = alice.makeGroup("Deux", joinedBy = listOf(bob))
        val todo = alice.makeTask(first.id, assignees = listOf(bob), dueAt = Fixed.dueA)
        val done = alice.makeTask(second.id, assignees = listOf(bob, alice))
        bob.tasks.setStatus(done.id, TaskStatus.DONE)
        val notMine = alice.makeTask(first.id, assignees = listOf(alice))
        val selfAssigned = bob.makeTask(first.id, assignees = listOf(bob))

        val open = bob.tasks.myTasks(includeDone = false)
        Verify.equal(open.map { it.id }.toSet(), setOf(todo.id, selfAssigned.id), "open tasks assigned to bob")
        val all = bob.tasks.myTasks(includeDone = true)
        Verify.equal(
            all.map { it.id }.toSet(), setOf(todo.id, done.id, selfAssigned.id), "all tasks assigned to bob",
        )
        Verify.that(all.none { it.id == notMine.id }, "tasks not assigned to bob are excluded")

        val names = mapOf(first.id to first.name, second.id to second.name)
        val assigners = mapOf(todo.id to alice.id, done.id to alice.id, selfAssigned.id to bob.id)
        for (item in all) {
            Verify.equal(item.groupName, names[item.groupId], "groupName of ${item.title}")
            Verify.that(item.myAssignedAt != null, "myAssignedAt of ${item.title}")
            Verify.equal(item.myAssignedBy, assigners[item.id], "myAssignedBy of ${item.title}")
            val plain = bob.tasks.task(item.id)
            Verify.equal(item.withoutPersonalFields, plain, "myTasks item ${item.title} matches task(id)")
        }
        val doneItem = Verify.unwrap(all.firstOrNull { it.id == done.id }, "done task in myTasks")
        Verify.equal(doneItem.assigneeIds, sortedIds(listOf(alice, bob)), "all assignees are listed, sorted")
        Verify.equal(doneItem.status, TaskStatus.DONE, "status of the done task")
        val aliceOpen = alice.tasks.myTasks(includeDone = false)
        Verify.equal(aliceOpen.map { it.id }.toSet(), setOf(notMine.id), "alice's open tasks")
    },

    ContractScenario("reads.assignmentsSince") { harness ->
        val alice = harness.user("Alice")
        val bob = harness.user("Bob")
        val group = alice.makeGroup(joinedBy = listOf(bob))
        val first = alice.makeTask(group.id, "Premiere", assignees = listOf(bob), dueAt = Fixed.dueA)
        val own = bob.makeTask(group.id, assignees = listOf(bob))
        val second = alice.makeTask(group.id, "Seconde", assignees = listOf(alice, bob))

        val events = bob.tasks.assignments(Fixed.longAgo)
        Verify.equal(events.map { it.taskId }, listOf(first.id, second.id), "assignments by others, oldest first")
        Verify.that(events.none { it.taskId == own.id }, "self-assignments are excluded")
        val event = events[0]
        Verify.equal(event.groupId, group.id, "event groupId")
        Verify.equal(event.taskTitle, first.title, "event taskTitle")
        Verify.equal(event.groupName, group.name, "event groupName")
        Verify.equal(event.assignedBy, alice.id, "event assignedBy")
        Verify.equal(event.dueAt, Fixed.dueA, "event dueAt")
        Verify.that(events[0].assignedAt < events[1].assignedAt, "events ordered by assignedAt")

        val later = bob.tasks.assignments(events[0].assignedAt)
        Verify.equal(later.map { it.taskId }, listOf(second.id), "since is exclusive")
        val none = bob.tasks.assignments(events[1].assignedAt)
        Verify.that(none.isEmpty(), "nothing after the last assignment, got $none")
        val aliceEvents = alice.tasks.assignments(Fixed.longAgo)
        Verify.that(aliceEvents.isEmpty(), "alice only assigned herself, got $aliceEvents")
    },
)
